src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs

```diff
@@ -1725,36 +1762,39 @@ public override IOperation VisitInvocation(IInvocationOperation operation, int?
-            if (!arguments.IsEmpty)
-            {
-                var builder = ArrayBuilder<IArgumentOperation>.GetInstance();
-                for (int i = arguments.Length - 1; i >= 0; i--)
-                {
-                    IOperation visitedValue = _evalStack.Pop();
-                    var originalArgument = (BaseArgument)arguments[i];
-                    builder.Add(new ArgumentOperation(visitedValue, originalArgument.ArgumentKind, originalArgument.Parameter,
-                                                      originalArgument.InConversionConvertibleOpt, originalArgument.OutConversionConvertibleOpt,
-                                                      semanticModel: null, originalArgument.Syntax, originalArgument.ConstantValue, originalArgument.IsImplicit));
-                }
-                builder.ReverseContents();
-                visitedArguments = builder.ToImmutableAndFree();
-            }
-            else
-            {
-                visitedArguments = ImmutableArray<IArgumentOperation>.Empty;
-            }
-
+            PushArray(arguments, arg => arg.Value);
+            ImmutableArray<IArgumentOperation> visitedArguments = PopArray(arguments.Length, RewriteArgumentFromArray(arguments));
             IOperation visitedInstance = operation.Instance == null ? null : _evalStack.Pop();
 
             return new InvocationExpression(operation.TargetMethod, visitedInstance, operation.IsVirtual, visitedArguments, semanticModel: null, operation.Syntax, operation.Type, operation.ConstantValue, operation.IsImplicit);
         }
 
+        public override IOperation VisitObjectCreation(IObjectCreationOperation operation, int? captureIdForResult)
+        {
+            ImmutableArray<IArgumentOperation> arguments = operation.Arguments;
+            PushArray(arguments, arg => arg.Value);
+            ImmutableArray<IArgumentOperation> visitedArgs = PopArray(arguments.Length, RewriteArgumentFromArray(arguments));
+
+            // Initializer is removed from the tree and turned into a series of statements that assign to the created instance
+            var objectCreation = new ObjectCreationExpression(operation.Constructor, initializer: null, visitedArgs, semanticModel: null, operation.Syntax, operation.Type, operation.ConstantValue, operation.IsImplicit);
+
+            // PROTOTYPE(dataflow): For a non-null initializer, we'll need to assign the created object to a flow reference and rewrite all initializers to assign to/call functions on it, then return return the flow reference
+
+            return objectCreation;
+        }
+
+        private Func<IOperation, int, IArgumentOperation> RewriteArgumentFromArray(ImmutableArray<IArgumentOperation> arguments)
+        {
+            return (visitedArgument, index) =>
+            {
+                Debug.Assert(index >= 0 && index < arguments.Length);
+                var originalArgument = (BaseArgument)arguments[index];
```

Comment:
>arguments [](start = 53, length = 9)

This creates a closure, I think it will be more efficient to pass this to the delegate as an additional argument

Reply:
Fixed.

---
In reply to: [171949387](https://github.com/dotnet/roslyn/pull/25178#discussion_r171949387) [](ancestors = 171949387)